src/lib/raft/raft.c

```diff
@@ -983,6 +1084,46 @@ raft_cfg_is_candidate(struct raft *raft, bool is_candidate)
 	raft_restore(raft);
 }
 
+bool
+raft_notify_is_leader_seen(struct raft *raft, bool is_leader_seen,
+			   uint32_t source)
+{
+	assert(source > 0 && source < VCLOCK_MAX && source != raft->self);
+	/* Leader doesn't care whether someone sees it or not. */
+	if (raft->state == RAFT_STATE_LEADER)
+		return false;
+
+	bool was_seen;
+	if (is_leader_seen) {
+		was_seen = bit_set(&raft->leader_witness_map, source);
+	} else {
+		was_seen = bit_clear(&raft->leader_witness_map, source);
+		if (was_seen)
+			raft_sm_election_update(raft);
+	}
+	return was_seen != is_leader_seen;
```

Comment:
You might want to simplify this now a little:
```Diff
diff --git a/src/lib/raft/raft.c b/src/lib/raft/raft.c
index d5f3a6113..607bfa9ab 100644
--- a/src/lib/raft/raft.c
+++ b/src/lib/raft/raft.c
@@ -1084,24 +1084,19 @@ raft_cfg_is_candidate(struct raft *raft, bool is_candidate)
 	raft_restore(raft);
 }
 
-bool
+void
 raft_notify_is_leader_seen(struct raft *raft, bool is_leader_seen,
 			   uint32_t source)
 {
 	assert(source > 0 && source < VCLOCK_MAX && source != raft->self);
 	/* Leader doesn't care whether someone sees it or not. */
 	if (raft->state == RAFT_STATE_LEADER)
-		return false;
-
-	bool was_seen;
-	if (is_leader_seen) {
-		was_seen = bit_set(&raft->leader_witness_map, source);
-	} else {
-		was_seen = bit_clear(&raft->leader_witness_map, source);
-		if (was_seen)
-			raft_sm_election_update(raft);
-	}
-	return was_seen != is_leader_seen;
+		return;
+	if (is_leader_seen)
+		bit_set(&raft->leader_witness_map, source);
+	else
+		bit_clear(&raft->leader_witness_map, source))
+	raft_sm_election_update(raft);
 }
 
 static bool
diff --git a/src/lib/raft/raft.h b/src/lib/raft/raft.h
index 314d2e014..773c11975 100644
--- a/src/lib/raft/raft.h
+++ b/src/lib/raft/raft.h
@@ -317,7 +317,7 @@ void
 raft_cfg_is_candidate(struct raft *raft, bool is_candidate);
 
 /** Notify this instance whether its peer sees the current leader. */
-bool
+void
 raft_notify_is_leader_seen(struct raft *raft, bool is_leader_seen,
 			   uint32_t source);
```

Reply:
I can't call `raft_sm_eleciton_update` regardless of `was_seen`. It's still not re-enterable, and I once again failed to make it re-enterable.
Imagine you start new elections, witness map is zero. Then you receive a confirmation that term was bumped from a remote node. This confirmation again contains `is_leader_seen = false` because there's no leader in the new term.
This will make the node call `raft_sm_election_update` once again and start new elections.

I made the return value `void` though.
```diff
diff --git a/src/lib/raft/raft.c b/src/lib/raft/raft.c
index 0fd77edb1..1f75a527f 100644
--- a/src/lib/raft/raft.c
+++ b/src/lib/raft/raft.c
@@ -408,24 +408,19 @@ raft_process_recovery(struct raft *raft, const struct raft_msg *req)
 	assert(!raft_is_enabled(raft));
 }
 
-bool
+void
 raft_notify_is_leader_seen(struct raft *raft, bool is_leader_seen,
 			   uint32_t source)
 {
 	assert(source > 0 && source < VCLOCK_MAX && source != raft->self);
 	/* Leader doesn't care whether someone sees it or not. */
 	if (raft->state == RAFT_STATE_LEADER)
-		return false;
+		return;
 
-	bool was_seen;
-	if (is_leader_seen) {
-		was_seen = bit_set(&raft->leader_witness_map, source);
-	} else {
-		was_seen = bit_clear(&raft->leader_witness_map, source);
-		if (was_seen)
-			raft_sm_election_update(raft);
-	}
-	return was_seen != is_leader_seen;
+	if (is_leader_seen)
+		bit_set(&raft->leader_witness_map, source);
+	else if (bit_clear(&raft->leader_witness_map, source))
+		raft_sm_election_update(raft);
 }
 
 /** Update raft state once leader is seen. */
diff --git a/src/lib/raft/raft.h b/src/lib/raft/raft.h
index 314d2e014..773c11975 100644
--- a/src/lib/raft/raft.h
+++ b/src/lib/raft/raft.h
@@ -317,7 +317,7 @@ void
 raft_cfg_is_candidate(struct raft *raft, bool is_candidate);
 
 /** Notify this instance whether its peer sees the current leader. */
-bool
+void
 raft_notify_is_leader_seen(struct raft *raft, bool is_leader_seen,
 			   uint32_t source);
 

```